src/VRCWrapper.ts

```diff
@@ -0,0 +1,214 @@
+        if (cookies === null) throw new Error('No cookies were set.');
+
+        await this.instanceCookie.parseCookieString(cookies, VRCWrapper.baseDomain);
+
+        // We save cookies if the user wants to keep them
+        if (process.env.USE_COOKIES === "true") {
+          try {
+            console.log("saving cookies");
+
+            await this.instanceCookie.saveCookies();
+          } catch (error) {
+            console.log("error saving cookies");
+
+            console.log("error: ", error);
+
+            if (error instanceof CookiesWriteError) {
+              throw new CookiesWriteError(`Error writing cookies.` + error.message);
+            } else if (error instanceof CookiesReadError) {
+              throw new CookiesReadError(`Error reading cookies.` + error.message);
+            } else if (error instanceof Error) {
+              throw new Error(`Unknown error writing cookies: ` + error.message);
+            }
+          }
+        }
+
+      }
+
+      const data = await response.json();
+
+      if ('success' in data) return;
+
+      try {
+        this.isAuthentificated = true;
+        const getCurrentUser = await this.authApi.getCurrentUser();
+        console.log(`${C.green}Logged in as: ${C.r}`, getCurrentUser.username);
+
+      } catch (error) {
+        this.isAuthentificated = false;
+      }
+
+      if (data.verified) {
+        throw new Error("2FA is already verified!");
+      }
+
+      // Handle special cases where authentication was not successful
+      if (data.requiresTwoFactorAuth) {
+        let isVerified = false;
+        if (JSON.stringify(data.requiresTwoFactorAuth) === JSON.stringify(["totp", "otp"])) {
+
+          console.log('Handling TOTP or OTP two-factor authentication...');
+          try {
+            const check2FA = await this.authApi.verify2FACodeTOTP({});
+            if (check2FA.verified) isVerified = true;
+            console.log("check2FA: ", check2FA);
+          } catch (error) {
+            console.log("error: ", error);
+          }
+
+
+        } else if (JSON.stringify(data.requiresTwoFactorAuth) === JSON.stringify(['emailOtp'])) {
+
+          console.log('Handling Email OTP two-factor authentication...');
+          try {
+            const check2FAEmail = await this.authApi.verify2FAEmailCode({});
+            if (check2FAEmail.verified) isVerified = true;
+          } catch (error) {
+            console.log("error: ", error);
+          }
+
+
+        } else {
+          // Handle other unknown two-factor authentication methods
+          throw new Error('Unknown error while authenticating!');
+        }
+
+        if (isVerified) {
+          console.log("2FA verified!");
+          this.isAuthentificated = true;
+        } else {
+          console.log("2FA not verified!");
+          this.isAuthentificated = false;
+        }
+
+      }
+
+
+    } catch (error) {
+      if (error instanceof TOTPRequired) {
+        error.logError();
+      }
```

Comment:
If you throw the error and the error has a log in their constructor you should not need this.